File: control/catalog.js
import { LOBBY } from './protocol.js';
import { safeSlot } from './tenancy.js';

export const MAX_SLOTS = 60;
const limits = { characterName: 80, race: 40, oath: 60, origin: 60, location: 100, playtime: 32, lastPlayed: 32 };
const regions = { eastluminant: 'EastLuminant', easternluminant: 'EastLuminant', theeasternluminant: 'EastLuminant',
  etreanluminant: 'EtreanLuminant', theetreanluminant: 'EtreanLuminant' };
export function menuRegion(value) { return typeof value === 'string' ? regions[value.toLowerCase().replace(/[\s_-]/g, '')] || null : null; }
const encoder = new TextEncoder();
export function cleanCatalog(input, accountId, at) {
  if (!input || input.type !== 'catalog' || input.version !== 1 || input.accountId !== accountId
      || input.placeId !== LOBBY || typeof input.complete !== 'boolean' || !Array.isArray(input.cards)
      || input.cards.length > MAX_SLOTS) return null;
  const entries = {}, seen = new Set();
  for (const item of input.cards) {
    if (!item || !safeSlot(item.slot) || !/^[A-Z]{1,3}$/.test(item.slot) || seen.has(item.slot)
        || typeof item.complete !== 'boolean') return null;
    seen.add(item.slot);
    const card = { slot: item.slot, source: 'menu-card', confirmed: false, observedAt: at };
    for (const [field, limit] of Object.entries(limits)) {
      const value = item[field];
      if (value == null) continue;
      if (typeof value !== 'string' || /[\x00-\x1f]/.test(value) || encoder.encode(value).length > limit) return null;
      card[field] = value;
    }
    if (item.level != null && (!Number.isInteger(item.level) || item.level < 0 || item.level > 1000)) return null;
    card.level = item.level ?? null;
    card.region = menuRegion(card.location);
    card.complete = item.complete && item.slotLabel === item.slot && !!card.characterName && !!card.race
      && !!card.location && card.level !== null;
    entries[card.slot] = card;
  }
  return { entries, complete: input.complete && seen.size > 0 && Object.values(entries).every(c => c.complete), observedAt: at };
}
export function selectable(entry, at) {
  return entry && (entry.confirmed === true || (entry.source === 'menu-card' && entry.complete === true))
    && Number.isInteger(entry.observedAt) && entry.observedAt <= at && at - entry.observedAt <= 86400;
}
const escape = value => String(value).replace(/[\\`*_~|<>@\[\]()]/g, '\\$&');
export function catalogPage(member, at, page = 1) {
  const entries = Object.values(member.slots || {}).sort((a, b) => a.slot.length - b.slot.length || a.slot.localeCompare(b.slot));
  const lines = entries.map(s => {
    const permission = member.approvedSlots?.[s.slot] ? 'approved' : 'not approved';
    const age = selectable(s, at) ? (s.confirmed ? 'verified in-world' : 'menu observed') : 'stale/incomplete';
    return `${escape(s.slot)}: ${escape(s.region || s.location || 'unsupported region')} | ${permission} | ${age}`
      + (s.characterName ? `\n${escape(s.characterName)} · Lv. ${s.level ?? '?'} ${escape(s.race || '?')}` : '')
      + ([s.oath, s.origin].filter(Boolean).length ? ` · ${[s.oath, s.origin].filter(Boolean).map(escape).join(' · ')}` : '')
      + (s.playtime || s.lastPlayed ? `\nPlayed ${escape(s.playtime || '?')} · Last ${escape(s.lastPlayed || '?')}` : '');
  });
  const groups = [[]]; let length = 0;
  for (const line of lines) {
    if (groups.at(-1).length >= 5 || length + line.length + 2 > 1750) { groups.push([]); length = 0; }
    groups.at(-1).push(line); length += line.length + 2;
  }
  if (!Number.isInteger(page) || page < 1 || page > groups.length) return `Choose page 1–${groups.length}.`;
  return `Page ${page}/${groups.length}\n${groups[page - 1].join('\n\n') || 'No characters yet. Run the paired loader at character selection.'}`;
}

File: control/catalog.lua
-- Compact menu metadata only; never send raw UI labels, friends, or local report files.
local Catalog = { VERSION = 1, MAX_CARDS = 60 }
local fields = { characterName = 80, race = 40, oath = 60, origin = 60, location = 100,
    playtime = 32, lastPlayed = 32 }
local function text(value, limit)
    if type(value) ~= "string" then return nil end
    value = value:gsub("[%z\1-\31]", " ")
    if not utf8.len(value) then return nil end
    if #value > limit then
        local ok, offset = pcall(utf8.offset, value, 0, limit + 1)
        value = value:sub(1, ok and offset and offset - 1 or limit)
    end
    if value == "" then return nil end
    return value
end
function Catalog.fromScan(report, accountId, at)
    if type(report) ~= "table" or type(report.cards) ~= "table" then return nil end
    local packet = { type = "catalog", version = 1, accountId = accountId,
        placeId = 4111023553, complete = not report.truncated and #report.cards > 0,
        cards = {}, status = report.status, observedAt = at }
    for _, source in ipairs(report.cards) do
        if #packet.cards >= Catalog.MAX_CARDS then packet.complete = false; break end
        if type(source.slot) == "string" and source.slot:match("^[A-Z]+$") and #source.slot <= 3 then
            local card = { slot = source.slot, slotLabel = source.slotLabel, level = source.level,
                complete = source.complete == true and not source.truncated and source.slotLabel == source.slot,
                confirmed = false, source = "menu-card", observedAt = at }
            for field, limit in pairs(fields) do card[field] = text(source[field], limit) end
            card.complete = card.complete and card.characterName ~= nil and card.race ~= nil and card.location ~= nil
                and type(card.level) == "number" and card.level % 1 == 0 and card.level >= 0 and card.level <= 1000
            packet.cards[#packet.cards + 1] = card
            if not card.complete then packet.complete = false end
        else packet.complete = false end
    end
    return packet
end
function Catalog.signature(packet)
    local ordered = table.clone(packet.cards)
    table.sort(ordered, function(a, b) return a.slot < b.slot end)
    local parts = { tostring(packet.complete) }
    for _, card in ipairs(ordered) do
        for _, field in ipairs({ "slot", "slotLabel", "complete", "level", "characterName", "race", "oath", "origin", "location", "playtime", "lastPlayed" }) do
            local value = tostring(card[field]); parts[#parts + 1] = #value .. ":" .. value
        end
    end
    return table.concat(parts, "|")
end
return Catalog
